contracts/psp22/lib.rs

```diff
@@ -0,0 +1,200 @@
+
+        fn _transfer_from_to(
+            &mut self,
+            from: &AccountId,
+            to: &AccountId,
+            value: Balance,
+            _data: Vec<u8>,
+        ) -> Result<(), PSP22Error> {
+            let from_balance = self.balance_of(*from);
+            if from_balance < value {
+                return Err(PSP22Error::InsufficientBalance);
+            }
+
+            self.balances.insert(from, &(from_balance - value));
+            let to_balance = self.balance_of(*to);
+            self.balances.insert(to, &(to_balance + value));
+
+            Self::emit_event(
+                self.env(),
+                Event::Transfer(Transfer {
+                    from: *from,
+                    to: *to,
+                    value,
+                }),
+            );
+
+            Ok(())
+        }
+
+        fn emit_event<EE>(emitter: EE, event: Event)
+        where
+            EE: EmitEvent<Self>,
+        {
+            emitter.emit_event(event);
+        }
+    }
+
+    impl PSP22 for Token {
+        /// Returns the total token supply.
+        #[ink(message)]
+        fn total_supply(&self) -> Balance {
+            self.total_supply
+        }
+
+        /// Returns the account balance for the specified `owner`.
+        #[ink(message)]
+        fn balance_of(&self, owner: AccountId) -> Balance {
+            self.balances.get(owner).unwrap_or_default()
+        }
+
+        /// Returns the amount which `spender` is allowed to withdraw on behalf of the `owner` account.
+        #[ink(message)]
+        fn allowance(&self, owner: AccountId, spender: AccountId) -> Balance {
+            self.allowances.get((owner, spender)).unwrap_or_default()
+        }
+
+        /// Allows `spender` to withdraw from the caller's account multiple times, up to the `value` amount.
+        #[ink(message)]
+        fn approve(&mut self, spender: AccountId, value: Balance) -> Result<(), PSP22Error> {
+            let owner = self.env().caller();
+            self._approve_from_to(owner, spender, value)
+        }
+
+        /// Incrase `spender`'s allowance to withdraw from the caller's account by the `by` amount.
+        #[ink(message)]
+        fn increase_allowance(
+            &mut self,
+            spender: AccountId,
+            by: Balance,
+        ) -> Result<(), PSP22Error> {
+            let owner = Self::env().caller();
+            self._approve_from_to(owner, spender, self.allowance(owner, spender) + by)
+        }
+
+        /// Decrease `spender`'s allowance to withdraw from the caller's account by the `by` amount.
+        #[ink(message)]
+        fn decrease_allowance(
+            &mut self,
+            spender: AccountId,
+            by: Balance,
+        ) -> Result<(), PSP22Error> {
+            let owner = Self::env().caller();
+            let allowance = self.allowance(owner, spender);
+
+            if allowance < by {
+                return Err(PSP22Error::InsufficientAllowance);
+            }
```

Comment:
we don't have symmetric check in `increase_allowance` - I would either have both or none